containers/server-postgres-image/uyuni-postgres-config.sh

```diff
@@ -0,0 +1,93 @@
+#!/bin/bash
+# Adjust postgresql.conf for Uyuni
+# This is run automatically at the first start of the container
+# or it can be run manually later.
+
+POSTGRESQL=/var/lib/pgsql/data/postgresql.conf
+SSL_CERT=/etc/pki/tls/certs/spacewalk.crt
+SSL_KEY=/etc/pki/tls/private/pg-spacewalk.key
+
+postgres_reconfig() {
+    echo "Setting $1 = $2"
+    if grep -E "^$1[[:space:]]*=" $POSTGRESQL >/dev/null; then
+        sed -i "s|^$1[[:space:]]*=.*|$1 = $2|" $POSTGRESQL
+    else
+        echo "$1 = $2" >> $POSTGRESQL
+    fi
+}
+
+# Get total memory in KB
+TOTAL_MEM_KB=$(grep MemTotal: /proc/meminfo | sed -e 's|MemTotal:[[:space:]]*\([0-9]*\).*|\1|')
+
+# Check minimum memory requirement (255KB)
+if [ "$TOTAL_MEM_KB" -lt $((0xff * 1024)) ]; then
+    echo "WARNING: low memory: $TOTAL_MEM_KB"
+    TOTAL_MEM_KB=$((0xff * 1024))
+fi
+
+# Binary rounding function
+bin_rnd() {
+    local value=$1
+    local mbt=1
+    while [ $value -gt 16 ]; do
+        value=$((value / 2))
+        mbt=$((mbt * 2))
+    done
+    echo $((mbt * value))
+}
+
+# Convert to MB
+to_mb() {
+    echo "$(($1 / 1024))MB"
+}
+
+# Get max_connections from current postgresql.conf
+MAX_CONNECTIONS=$(grep -E "^max_connections[[:space:]]*=" $POSTGRESQL | sed -e 's/.*=[[:space:]]*\([0-9]*\).*/\1/')
+[ -z "$MAX_CONNECTIONS" -o "$MAX_CONNECTIONS" -lt 400 ] && MAX_CONNECTIONS=500
+
+# Check if system uses SSD
+if [ "$(cat /sys/block/$(mount | grep " /var/lib/pgsql/data " | cut -d' ' -f1 |sed -e 's/[0-9][0-9p]*//g' -e 's|/dev/||' )/queue/rotational)" -gt 0 ]; then
+    IS_SSD=0
+    echo "Rotational HDD detected."
+else
+    IS_SSD=1
+    echo "SSD detected."
+fi
+
+# Calculate values
+SHARED_BUFFERS=$(bin_rnd $((TOTAL_MEM_KB / 4)))
+EFFECTIVE_CACHE_SIZE=$(bin_rnd $((TOTAL_MEM_KB * 3 / 4)))
+WORK_MEM=$(bin_rnd $(((TOTAL_MEM_KB - SHARED_BUFFERS) / (3 * MAX_CONNECTIONS))))
+MAINTENANCE_WORK_MEM=$(bin_rnd $(( TOTAL_MEM_KB / 16 < 1048576 ? TOTAL_MEM_KB / 16 : 1048576 ))) # 1GB
+
+# Apply configurations
+postgres_reconfig "shared_buffers" "$(to_mb $SHARED_BUFFERS)"
+postgres_reconfig "effective_cache_size" "$(to_mb $EFFECTIVE_CACHE_SIZE)"
+postgres_reconfig "work_mem" "$(to_mb $WORK_MEM)"
+postgres_reconfig "maintenance_work_mem" "$(to_mb $MAINTENANCE_WORK_MEM)"
+postgres_reconfig "max_wal_size" "4096MB"
+postgres_reconfig "min_wal_size" "1024MB"
+postgres_reconfig "checkpoint_completion_target" "0.9"
+postgres_reconfig "wal_buffers" "16MB"
+postgres_reconfig "constraint_exclusion" "off"
+postgres_reconfig "max_connections" "$MAX_CONNECTIONS"
+
+if [ "$IS_SSD" -eq 1 ]; then
+    postgres_reconfig "random_page_cost" "1.1"
+    postgres_reconfig "effective_io_concurrency" "200"
+else
+    postgres_reconfig "random_page_cost" "4"
+    postgres_reconfig "effective_io_concurrency" "2"
+fi
```

Comment:
Related to the discussion above.

Personally I would get rid of this check and use

```suggestion
if [ "$IS_SSD" -eq 1 ]; then
postgres_reconfig "random_page_cost" "1.1"
postgres_reconfig "effective_io_concurrency" "20"
```

BTW. `random_page_cost` 1.1 is recommended by postgresql config for SSD disks and this value influences how likely planner is to select index scan instead of sequential scan.
Default sequential scan cost is 1.

BTW 2. Both of these options can also be overridden per table.

Reply:
agree, move forward with ssd only.